File: py/src/jsonrpc3/http_server.py
"""HTTP server for JSON-RPC 3.0.

Note: This HTTP server uses Python's built-in http.server which is synchronous.
WebSocket support requires an async framework. For production use with WebSockets,
consider using an ASGI server like uvicorn with async handlers.

This implementation detects WebSocket upgrade attempts and returns a helpful error.
"""

import http.server
import socketserver
import threading
from typing import Optional

from .encoding import get_codec
from .errors import RpcError
from .handler import Handler
from .types import (
    MIME_TYPE_JSON,
    MessageSet,
    batch_response_to_message_set,
    is_request,
    new_error_response,
    to_batch,
    to_request,
)


class HttpServer:
    """HTTP server for JSON-RPC 3.0 using http.server."""

    def __init__(
        self,
        handler: Handler,
        port: int = 3000,
        host: str = "localhost",
        mime_type: str = MIME_TYPE_JSON,
    ):
        """
        Initialize HTTP server.

        Args:
            handler: Request handler
            port: Port to listen on (use 0 for random available port)
            host: Host to bind to
            mime_type: MIME type for encoding
        """
        self.handler = handler
        self._requested_port = port
        self._host = host
        self._codec = get_codec(mime_type)
        self._server: Optional[socketserver.TCPServer] = None
        self._server_thread: Optional[threading.Thread] = None
        self.port: Optional[int] = None

    def start(self) -> "HttpServer":
        """Start the HTTP server."""

        # Create request handler class with access to our handler
        handler_instance = self.handler
        codec_instance = self._codec

        class RequestHandler(http.server.BaseHTTPRequestHandler):
            """Custom request handler for JSON-RPC."""

            def log_message(self, format, *args):
                """Suppress default logging."""
                pass

            def do_POST(self):
                """Handle POST requests."""
                try:
                    # Read request body
                    content_length = int(self.headers.get("Content-Length", 0))
                    body_bytes = self.rfile.read(content_length)

                    # Decode message set
                    msg_set = codec_instance.unmarshal_messages(body_bytes)

                    # Handle batch or single request
                    if msg_set.is_batch:
                        batch = to_batch(msg_set)
                        batch_resp = handler_instance.handle_batch(batch)
                        response_msg_set = batch_response_to_message_set(batch_resp)
                    else:
                        request = to_request(msg_set)
                        response = handler_instance.handle_request(request)
                        response_msg_set = MessageSet(
                            messages=[response], is_batch=False
                        )

                    # Encode response
                    resp_bytes = codec_instance.marshal_messages(response_msg_set)

                    # Send response
                    self.send_response(200)
                    self.send_header("Content-Type", codec_instance.mime_type)
                    self.send_header("Content-Length", str(len(resp_bytes)))
                    self.end_headers()
                    self.wfile.write(resp_bytes)

                except RpcError as e:
                    # Return RPC error as valid response
                    error_resp = new_error_response(e.to_dict(), None)
                    resp_msg_set = MessageSet(
                        messages=[error_resp], is_batch=False
                    )
                    resp_bytes = codec_instance.marshal_messages(resp_msg_set)

                    self.send_response(200)
                    self.send_header("Content-Type", codec_instance.mime_type)
                    self.send_header("Content-Length", str(len(resp_bytes)))
                    self.end_headers()
                    self.wfile.write(resp_bytes)

                except Exception as e:
                    # Internal server error
                    error_msg = f"Internal Server Error: {str(e)}"
                    self.send_error(500, error_msg)

            def do_GET(self):
                """Handle GET requests - check for WebSocket upgrade."""
                # Check if this is a WebSocket upgrade request
                upgrade = self.headers.get("Upgrade", "").lower()
                connection = self.headers.get("Connection", "").lower()

                if upgrade == "websocket" and "upgrade" in connection:
                    # WebSocket upgrade detected - return helpful error
                    self.send_response(501)
                    self.send_header("Content-Type", "text/plain")
                    self.end_headers()
                    msg = (
                        "WebSocket upgrade detected. This HTTP server is synchronous and does not support WebSocket upgrades.\n"
                        "For WebSocket support, use the WebSocketHandler with an async server, or use an ASGI server like uvicorn."
                    )
                    self.wfile.write(msg.encode())
                else:
                    # Regular GET request - not allowed
                    self.send_error(405, "Method Not Allowed")

        # Create threaded HTTP server
        class ThreadedHTTPServer(socketserver.ThreadingMixIn, http.server.HTTPServer):
            """HTTP server with threading support."""

            daemon_threads = True

        # Start server in background thread
        def run_server():
            self._server = ThreadedHTTPServer((self._host, self._requested_port), RequestHandler)
            self.port = self._server.server_address[1]
            self._server.serve_forever()

        self._server_thread = threading.Thread(target=run_server, daemon=True)
        self._server_thread.start()

        # Wait for server to start
        import time
        while self.port is None:
            time.sleep(0.01)

        return self

    def stop(self) -> None:
        """Stop the HTTP server."""
        if self._server:
            self._server.shutdown()
            self._server.server_close()
            self._server = None

        if self._server_thread:
            self._server_thread.join(timeout=1.0)
            self._server_thread = None

        self.port = None

    @property
    def url(self) -> Optional[str]:
        """Get server URL."""
        if self._server and self.port:
            return f"http://{self._host}:{self.port}"
        return None

File: ruby/lib/jsonrpc3/http_server.rb
# frozen_string_literal: true

require "rack"
require "puma"
require "faye/websocket"

module JSONRPC3
  # HTTP Server for JSON-RPC 3.0 using Rack
  # Automatically upgrades to WebSocket when upgrade headers are detected
  class HttpServer
    attr_reader :handler, :port

    def initialize(root_object, options = {})
      mime_type = options[:mime_type] || MIME_TYPE_JSON
      @codec = JSONRPC3.get_codec(mime_type)
      # HTTP requests don't support callbacks, so use NoOpCaller
      session = Session.new
      @handler = Handler.new(session, root_object, NoOpCaller.new, [mime_type])
      @requested_port = options[:port] || 3000
      @host = options[:host] || "localhost"
      @server = nil
      @server_thread = nil
      @port = nil
    end

    # Rack call interface
    def call(env)
      # Check for WebSocket upgrade request
      if Faye::WebSocket.websocket?(env)
        return handle_websocket_upgrade(env)
      end

      # Only accept POST requests for JSON-RPC
      unless env["REQUEST_METHOD"] == "POST"
        return [405, { "Content-Type" => "text/plain" }, ["Method Not Allowed"]]
      end

      # Read request body
      body_bytes = env["rack.input"].read

      begin
        # Decode message set
        msg_set = @codec.unmarshal_messages(body_bytes)

        # Handle batch or single request
        response_msg_set = if msg_set.is_batch
                             batch = JSONRPC3.to_batch(msg_set)
                             batch_resp = @handler.handle_batch(batch)
                             JSONRPC3.batch_response_to_message_set(batch_resp)
                           else
                             request = JSONRPC3.to_request(msg_set)
                             response = @handler.handle_request(request)
                             MessageSet.new(messages: [response], is_batch: false)
                           end

        # Encode response
        resp_bytes = @codec.marshal_messages(response_msg_set)

        [200, { "Content-Type" => @codec.mime_type }, [resp_bytes]]
      rescue RpcError => e
        # Return RPC error as valid response
        error_resp = JSONRPC3.new_error_response(e.to_h, nil)
        resp_msg_set = MessageSet.new(messages: [error_resp], is_batch: false)
        resp_bytes = @codec.marshal_messages(resp_msg_set)
        [200, { "Content-Type" => @codec.mime_type }, [resp_bytes]]
      rescue StandardError => e
        # Internal server error
        [500, { "Content-Type" => "text/plain" }, ["Internal Server Error: #{e.message}"]]
      end
    end

    # Start the server
    def start
      @server_thread = Thread.new do
        @server = Puma::Server.new(self)
        @server.add_tcp_listener(@host, @requested_port)
        @server.run
      end

      # Wait for server to start and get actual port
      sleep 0.1 while @server.nil?

      # Get the actual bound port
      if @server
        @port = @server.connected_ports.first || @requested_port
      end

      self
    end

    # Stop the server
    def stop
      @server&.stop(true)
      @server_thread&.join
      @server = nil
      @server_thread = nil
    end

    # Get server URL
    def url
      @server && @port ? "http://#{@host}:#{@port}" : nil
    end

    private

    # Handle WebSocket upgrade request
    # This allows the HTTP server to seamlessly support both HTTP JSON-RPC
    # and WebSocket JSON-RPC without requiring separate handlers
    def handle_websocket_upgrade(env)
      # Upgrade to WebSocket with jsonrpc3.json subprotocol
      ws = Faye::WebSocket.new(env, ["jsonrpc3.json"])

      # Only JSON is supported for now
      content_type = MIME_TYPE_JSON

      # Reject non-JSON protocols
      if ws.protocol && ws.protocol != "jsonrpc3.json"
        ws.close(1002, "Only jsonrpc3.json protocol is supported")
        return ws.rack_response
      end

      # Create connection handler - use the same root object from HTTP server
      conn = WebSocketServerConn.new(ws, @handler.root_object, content_type, [@codec.mime_type])

      # Set up event handlers
      ws.on :open do |event|
        conn.on_open
      end

      ws.on :message do |event|
        conn.on_message(event.data)
      end

      ws.on :close do |event|
        conn.on_close(event.code, event.reason)
      end

      ws.on :error do |event|
        conn.on_error(event.message)
      end

      # Return async Rack response
      ws.rack_response
    end
  end
end
